### ISSUE TITLE
Consider removing SymbolInformation.owner from the schema

### ISSUE THREAD

**Issue by xeno-by** [scalameta/scalameta#1540]
Now that we have a symbol parser in `semanticdb3`, `SymbolInformation.owner` seems redundant. We should consider removing it from the schema to avoid confusion and potential discrepancies with updates to the implementation of the symbol parser.

**Comment by xeno-by** [comment]
Unfortunately, this doesn't work. `ScalaSymbolOps.owner` only returns correct results for global symbols. Local members, e.g. members of local classes, must obtain their owner information elsewhere.

**Comment by xeno-by** [comment]
I have experimented with this in https://github.com/xeno-by/scalameta/commit/f435f58b347f17c9c3c9c49218427644c63c1125 and must say that this is not a clear-cut win. Scalac's notion of owners is pretty crazy, so we'd have to clean it up and spec before readding SymbolInformation.owner. Since this doesn't look critical for our work in 4.0.0, I'm removing the milestone assignment.

**Comment by olafurpg** [comment]
One situation where we may need `SymbolInformation.owner` is to handle static inner java classes. The proposed solution for #1491 is to give the symbol syntax `java.util.Map.Entry#` for the inner class `Entry` while it's owner should be `java.util.Map#`.

**Comment by xeno-by** [comment]
The argument that @olafurpg has mentioned above is no longer applicable. In https://github.com/scalameta/scalameta/pull/1638, we went the other way and retained syntactic owners for global symbols.

**Comment by xeno-by** [comment]
I'm working on pickling deeply nested structural/existential types, and I can attest that the lack of `SymbolInformation.owner` is pretty annoying. I'm left with no choice but to maintain an owner chain while I'm traversing SemanticDB payloads. Perhaps, we could reintroduce `owner`, but only fill it in for local symbols?

**Comment by olafurpg** [comment]
I have experienced the same problem, I am in favor of adding `.owner` but only fill it for local symbols 👍

**Comment by olafurpg** [comment]
Encountering the lack of `owner` when implementing rename in Metals for anonymous classes. A workaround is to filter out class symbols that contain the local method symbol as part of `ClassSignature.declarations`.

**Comment by jozanek** [comment]
Metals recovers a local symbol's owner by scanning all `SymbolInformation`s for a `ClassSignature.declarations` containing it, and `RenameProvider` refuses local classes outright.

Concrete proposal (per the consensus above): reintroduce owner **only for local symbols**:

```protobuf
string owner = 21;
```

- Empty for globals (derivable from the symbol string); owner of a local = nearest enclosing definition *that has a symbol* (may itself be local, e.g. members of `$anon`).
- Inverse of `ClassSignature.declarations` where both exist; hardlinks don't fill it, keeping metac/metacp in agreement (what sank xeno-by/scalameta@f435f58).
- Additive field -> minor release, same as `overridden_symbols`/`documentation`.

Scala 3 `ExtractSemanticDB` and mtags would need to adopt the field too.

@tgodzik still worth having at the SemanticDB level? If yes, I'll open both PRs.

**Comment by tgodzik** [comment]
Do we have a case that doesn't work with renames currently? I would rather avoid adding a new field if we can handle it in the local context (get all information from the current semantcdb file)

**Comment by jozanek** [comment]
@tgodzik  you're right. I rechecked and nothing fails: anon-override rename works (tested in `RenameLspSuite`), since locals never cross files the `declarations` scan always succeeds. What's left is just avoiding an O(#symbols) scan, not worth a schema field. Feel free to close this issue.
